polygeist/tools/cgeist/Lib/CGExpr.cc

```diff
@@ -2901,3 +2903,76 @@ ValueCategory MLIRScanner::EmitBinOr(const BinOpInfo &Info) {
                                              Info.getRHS().getValue(builder)),
                        /*isReference*/ false);
 }
+
+#define HANDLEUNARYOP(OP)                                                      \
+  ValueCategory MLIRScanner::VisitUnary##OP(UnaryOperator *E,                  \
+                                            QualType PromotionType) {          \
+    LLVM_DEBUG({                                                               \
+      llvm::dbgs() << "VisitUnary" #OP ": ";                                   \
+      E->dump();                                                               \
+      llvm::dbgs() << "\n";                                                    \
+    });                                                                        \
+    QualType promotionTy =                                                     \
+        PromotionType.isNull()                                                 \
+            ? Glob.getTypes().getPromotionType(E->getSubExpr()->getType())     \
+            : PromotionType;                                                   \
+    auto result = Visit##OP(E, promotionTy);                                   \
+    if (result.val && !promotionTy.isNull())                                   \
+      result = EmitUnPromotedValue(getMLIRLocation(E->getExprLoc()), result,   \
+                                   E->getType());                              \
+    return result;                                                             \
+  }
+#include "Expressions.def"
+
```

Comment:
Add #undef HANDLEUNARYOP

Reply:
Answered above

Reply:
I'd prefer the undef to be outside the definition file. Clearer IMO.